<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2
import warnings
warnings.filterwarnings('ignore')

In [2]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [3]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies).

In [4]:
# текст запроса
query_3_1 = f'''
select count(id)
from public.vacancies
'''

In [5]:
# результат запроса
df31 = pd.read_sql_query(query_3_1, connection)
display(df31)  # 49197

count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers).

In [6]:
# текст запроса
query_3_2 = f'''
select count(id)
from public.employers
'''

In [7]:
# результат запроса
df32 = pd.read_sql_query(query_3_2, connection)
display(df32)  # 23501

count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [8]:
# текст запроса
query_3_3 = f'''
select count(id)
from public.areas
'''

In [9]:
# результат запроса
df33 = pd.read_sql_query(query_3_3, connection)
display(df33)  # 1362

count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
# текст запроса
query_3_4 = f'''
select count(id)
from public.industries
'''

In [11]:
# результат запроса
df34 = pd.read_sql_query(query_3_4, connection)
display(df34)  # 294

count
0    294

## Выводы по предварительному анализу данных

В результате предварительного анализа данных выяснилось, что в базе данных содержится 23501 работодателей из 1362 регионов, представляющие 294 сферы деятельности, которые выложили 49197 вакансий.

---

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [12]:
# текст запроса
query_4_1 = f'''
select  a.name area,
        count(v.id) cnt
from public.vacancies v
  join public.areas a on v.area_id = a.id
group by a.name
order by cnt desc
limit 5
'''

In [13]:
# результат запроса
df41 = pd.read_sql_query(query_4_1, connection)
display(df41)

area   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [14]:
# текст запроса
query_4_2 = f'''
select  count(id)
from public.vacancies
where salary_from is not null
   or salary_to is not null
'''

In [15]:
# результат запроса
df42 = pd.read_sql_query(query_4_2, connection)
display(df42)  # 24073

count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [16]:
# текст запроса
query_4_3 = f'''
select  round(avg(salary_from)),
        round(avg(salary_to))
from public.vacancies
'''

In [17]:
# результат запроса
df43 = pd.read_sql_query(query_4_3, connection)
display(df43)  # 71065 110537

round     round
0  71065.0  110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [18]:
# текст запроса
query_4_4 = f'''
select  schedule,
        employment,
        count(id)
from public.vacancies
group by schedule, employment
order by 3 desc
'''

In [19]:
# результат запроса
df44 = pd.read_sql_query(query_4_4, connection)
display(df44)

schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта.

In [20]:
# текст запроса
query_4_5 = f'''
select  experience
from public.vacancies
group by experience
order by count(id)
'''

In [21]:
# результат запроса
df45 = pd.read_sql_query(query_4_5, connection)
display(df45)

experience
0         Более 6 лет
1           Нет опыта
2       От 3 до 6 лет
3  От 1 года до 3 лет

## Выводы по детальному анализу вакансий

В результате детального анализа вакансий можно сказать, что большая часть вакантных рабочих мест распределена по крупным городам. Около половины работодателей не указали минимальный, либо максимальный уровень предлагаемой оплаты труда, причем средний минимальный уровень зарплаты составляет около 71 тыс. рублей, а максимальный - 110 тыс. рублей.
Подавляющее большинство вакансий предполагают полную занятость при полном рабочем дне. Причем рынок ожидает в большей степени молодых специалистов с опытом работы от одного года до трёх лет.

---

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [22]:
# текст запроса
query_5_1 = f'''
select  e.name employer_name,
        count(v.id) vacancies_count
from public.employers e
  join public.vacancies v on e.id = v.employer_id
group by e.name
order by count(v.id) desc
limit 5
'''

In [23]:
# результат запроса
df51 = pd.read_sql_query(query_5_1, connection)
display(df51)

employer_name  vacancies_count
0         Яндекс             1933
1     Ростелеком              491
2       Тинькофф              444
3           СБЕР              428
4  Газпром нефть              331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [24]:
query_5_2 = f'''
select  a.name,
        count(distinct e.id) employers_count
from public.areas a
  left join public.employers e on a.id = e.area
  left join public.vacancies v on a.id = v.area_id
where v.id is null
group by a.name
order by employers_count desc
'''

In [25]:
# результат запроса
df52 = pd.read_sql_query(query_5_2, connection)
display(df52)

name  employers_count
0                            Россия              410
1                         Казахстан              207
2                Московская область               75
3                Краснодарский край               19
4                          Беларусь               18
..                              ...              ...
588                        Кувандык                0
589                        Куйбышев                0
590  Куйбышево (Ростовская область)                0
591                        Култаево                0
592                        Кульсары                0

[593 rows x 2 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [26]:
# текст запроса
query_5_3 = f'''
select  e.name vacancies_name,
        count(distinct v.area_id) area_count
from public.vacancies v
  join public.employers e on v.employer_id = e.id
group by e.name
order by area_count desc
'''

In [27]:
# результат запроса
df53 = pd.read_sql_query(query_5_3, connection)
display(df53)

vacancies_name  area_count
0                      Яндекс         181
1                  Ростелеком         152
2                  Спецремонт         116
3      Поляков Денис Иванович          88
4                   ООО ЕФИН           71
...                       ...         ...
14761                  UniSol           1
14762            UNISTORY LLC           1
14763                   UNIT6           1
14764     United Distribution           1
14765              UNITEL LLC           1

[14766 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности.

In [28]:
# текст запроса
query_5_4 = f'''
select count(e.id)
from public.employers e
  left join public.employers_industries ei on e.id = ei.employer_id
where ei.industry_id is null
'''

In [29]:
# результат запроса
df54 = pd.read_sql_query(query_5_4, connection)
display(df54)  # 8419

count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности.

In [30]:
# текст запроса
query_5_5 = f'''
select  e.name employers_name,
        count(distinct i.name) industries_name
from public.employers e
  join public.employers_industries ei on e.id = ei.employer_id
  join public.industries i on ei.industry_id = i.id
group by e.name
having count(distinct i.name) = 4
order by e.name
--offset 2 limit 1
'''

In [31]:
# результат запроса
df55 = pd.read_sql_query(query_5_5, connection)
display(df55)  # 2ГИС

employers_name  industries_name
0        101 Интернет                4
1            21vek.by                4
2                2ГИС                4
3                  2К                4
4         4 пикселя +                4
...               ...              ...
1127            ЮРИОН                4
1128  ЮТИП Технологии                4
1129           ЯКласс                4
1130            ЯрНео                4
1131    Яшин&Партнёры                4

[1132 rows x 2 columns]

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [32]:
# текст запроса
query_5_6 = f'''
select  count(e.id) employers_count,
        i.name
from public.employers e
  left join public.employers_industries ei on e.id = ei.employer_id
  left join public.industries i on ei.industry_id = i.id
where i.name = 'Разработка программного обеспечения'
group by i.name
'''

In [33]:
# результат запроса
df56 = pd.read_sql_query(query_5_6, connection)
display(df56)  # 3553

employers_count                                 name
0             3553  Разработка программного обеспечения

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8).

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API.

In [34]:
# код для получения списка городов-милионников
import requests
from bs4 import BeautifulSoup
url='https://ru.wikipedia.org/wiki/Города-миллионеры_России'
response=requests.get(url)
page=BeautifulSoup(response.text,'html.parser')

cities_here=str(page.find('table', class_="standard"))
cities_here=cities_here.split('title="')[3:]

cities=[]
for i in cities_here:
    cities.append(i.split('">')[0])

In [35]:
c = tuple(cities)

In [36]:
# текст запроса
query_5_7 = f'''
with t as (select  a.name,
        count(*) cnt
from public.employers e
  join public.vacancies v on e.id = v.employer_id
  join public.areas a on a.id = v.area_id
where e.name = 'Яндекс'
          and a.name in {c}
group by a.id
order by cnt)

select *
from t

union all

select  'Total',
        sum(cnt)
from t
'''

In [37]:
# результат запроса
df57 = pd.read_sql_query(query_5_7, connection)
display(df57)

name    cnt
0              Омск   21.0
1         Челябинск   22.0
2        Красноярск   23.0
3         Волгоград   24.0
4    Ростов-на-Дону   25.0
5            Казань   25.0
6             Пермь   25.0
7            Самара   26.0
8               Уфа   26.0
9         Краснодар   30.0
10          Воронеж   32.0
11      Новосибирск   35.0
12  Нижний Новгород   36.0
13     Екатеринбург   39.0
14  Санкт-Петербург   42.0
15           Москва   54.0
16            Total  485.0

## Выводы по анализу работодателей

Из всех работодателей на рынке больше всего вакансий предоставляют Яндекс, Ростелеком, СБЕР и Газпром нефть. Причем наибольшее количество работодателей ищут кандидатов в Казахстане и Московской области. 36% работодателей не указывают сферу своей деятельности. 15% работодателей занимаются разработкой ПО. Так компания Яндекс предлагает работу в городах-миллионниках, в топ которых входят Москва, Петербург и Екатеринбург.

***

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.*


In [38]:
# текст запроса
query_6_1 = f'''
select count(lower(name)) vacancy_count
from public.vacancies
where name ilike '%data%'
      or name ilike '%данн%'
'''

In [39]:
# результат запроса
df61 = pd.read_sql_query(query_6_1, connection)
display(df61)  # 1771

vacancy_count
0           1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста?
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.


In [40]:
# текст запроса
query_6_2 = f'''
select count(*)
from public.vacancies
where   (name ilike '%data scientist%'
      or name ilike '%data science%'
      or name ilike '%исследователь данных%'
      or (name like '%ML%' and name not ilike '%HTML%')
      or name ilike '%machine learning%'
      or name ilike '%машинн%обучен%')

    and   (experience ilike '%Нет опыта%'
          or name ilike '%junior%'
          or employment ilike '%Cтажировка%')
'''

In [41]:
# query_6_2 = f'''
# select
#     v.name
# from vacancies v
# where
#     (v.name ~* 'data scien(ce|tist)|исследователь данных|ML|machine learning|машинн.*обучен.*'
#         and v.name !~* 'HTML')
#     and (v.name ~* 'junior'
#           or v.experience = 'Нет опыта'
#           or v.employment = 'Стажировка')
# '''

In [42]:
# результат запроса
df62 = pd.read_sql_query(query_6_2, connection)
display(df62)  # 51

count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [43]:
# текст запроса
query_6_3 = f'''
select count(*)
  --name, key_skills
from public.vacancies
where   (name ilike '%data scientist%'
      or name ilike '%data science%'
      or name ilike '%исследователь данных%'
      or (name ilike '%ML%' and name not ilike '%HTML%')
      or name ilike '%machine learning%'
      or name ilike '%машинн%обучен%')

    and   ((key_skills ilike '%SQL%'
          or key_skills ilike '%postgres%')
          and key_skills is not null)
'''

In [44]:
# результат запроса
df63 = pd.read_sql_query(query_6_3, connection)
display(df63)  # 229

count
0    229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [45]:
# текст запроса
query_6_4 = f'''
select count(*)
  --name, key_skills
from public.vacancies
where   (name ilike '%data scientist%'
      or name ilike '%data science%'
      or name ilike '%исследователь данных%'
      or (name ilike '%ML%' and name not ilike '%HTML%')
      or name ilike '%machine learning%'
      or name ilike '%машинн%обучен%')

    and   ((key_skills ilike '%python%')
          and key_skills is not null)
'''

In [46]:
# результат запроса
df64 = pd.read_sql_query(query_6_4, connection)
display(df64)  # 357

count
0    357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [47]:
# текст запроса
query_6_5 = f'''
SELECT round(avg(length(key_skills) - length(replace(key_skills, CHR(9), '')) + 1), 2)
from public.vacancies
where key_skills is not null
    and  (name ilike '%data scientist%'
       or name ilike '%data science%'
       or name ilike '%исследователь данных%'
       or (name like '%ML%' and name not ilike '%HTML%')
       or name ilike '%machine learning%'
       or name ilike '%машинн%обучен%')
'''

In [48]:
# результат запроса
df65 = pd.read_sql_query(query_6_5, connection)
display(df65) # 6.41

round
0   6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*).

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос `select 1 + null`). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа.

In [49]:
# текст запроса
query_6_6 = f'''
select distinct experience, round(avg(coalesce((salary_from + salary_to)/2, salary_from, salary_to))) avg_salary
from public.vacancies
where   (name ilike '%data scientist%'
      or name ilike '%data science%'
      or name ilike '%исследователь данных%'
      or (name like '%ML%' and name not ilike '%HTML%')
      or name ilike '%machine learning%'
      or name ilike '%машинн%обучен%')
    and
      (salary_from is not null or salary_to is not null)
group by experience
'''

In [50]:
# результат запроса
df66 = pd.read_sql_query(query_6_6, connection)
display(df66)  # 243115

experience  avg_salary
0           Нет опыта     74643.0
1  От 1 года до 3 лет    139675.0
2       От 3 до 6 лет    243115.0

***

## Выводы по предметному анализу

В этом юните мы исследовали вакансии относящиеся к DS.

Всего вакансий, относящихся к работе с данными нашлось 1771, что составляет 3.6%  от всех записей в базе данных.

Вакансий, относящихся к DS - 480, что составляет 27% от всего количества вакансий, относящихся к работе с данными.

11% вакансий DS ожидают специалистов уровня Junior.

48% вакансий DS требуют знания SQL или Postgres.

74% вакансий DS требуют знания Python.

В среднем вакансии для DS требуют 6 ключевых навыков у соискателей.

В зависимост от опыта работодатели предлагают следующие усреднённые зарплаты:

Опыт               | Оплата труда
:-                 |-:
Нет опыта          |74643
От 1 года до 3 лет |139675
От 3 до 6 лет      |243115

# Общий вывод по проекту

In [51]:
cnt_vac = pd.read_sql_query(''' select count(*)
                                from public.vacancies''', connection).values[0][0]
cnt_unic_vac = pd.read_sql_query('''select count(distinct name)
                                    from public.vacancies''', connection).values[0][0]
top = pd.read_sql_query(''' select distinct name, count(id)
                            from public.vacancies
                            group by name
                            order by count(id) desc
                            limit 2''', connection)
top_city = pd.read_sql_query('''select
                                      a.name city,
                                      count(v.name) count_vacancy
                                from public.vacancies v
                                  join public.areas a on v.area_id = a.id
                                  join public.employers e on v.employer_id = e.id
                                group by a.name
                                order by count(v.name) desc
                                limit 5''', connection)
cnt_empl = pd.read_sql_query('''select
                                      a.name city,
                                      count(case when v.employment = 'Полная занятость'    then v.employment end) as "Полная занятость",
                                      count(case when v.employment = 'Частичная занятость' then v.employment end) as "Частичная занятость",
                                      count(case when v.employment = 'Стажировка'          then v.employment end) as "Стажировка",
                                      count(case when v.employment = 'Проектная работа'    then v.employment end) as "Проектная работа"
                                from public.vacancies v
                                  join public.areas a on v.area_id = a.id
                                  join public.employers e on v.employer_id = e.id
                                group by a.name
                                order by 2 desc''', connection)

In [52]:
print(f'''В базе данных содержится {cnt_vac} записей о вакансиях.

Из них {cnt_unic_vac} уникальных.

В топе находятся:''')
display(top)

В базе данных содержится 49197 записей о вакансиях.

Из них 21223 уникальных.

В топе находятся:


name  count
0  Системный администратор   1761
1           Программист 1С   1153

In [53]:
print(f'''По количеству вакансий можно выделить топ {len(top_city)}-ти городов:''')
display(top_city)

По количеству вакансий можно выделить топ 5-ти городов:


city  count_vacancy
0           Москва           5333
1  Санкт-Петербург           2851
2            Минск           2112
3      Новосибирск           2006
4           Алматы           1892

In [54]:
print(f'''В качестве дополнительных исследований можно вывести сводную таблицу количества типов трудоустройства по городам''')
display(cnt_empl)

В качестве дополнительных исследований можно вывести сводную таблицу количества типов трудоустройства по городам


city  Полная занятость  Частичная занятость  Стажировка  \
0              Москва              5118                   66         125   
1     Санкт-Петербург              2716                   59          63   
2               Минск              2011                   60          27   
3         Новосибирск              1945                   28          26   
4              Алматы              1819                   33          27   
..                ...               ...                  ...         ...   
764          Коммунар                 0                    1           0   
765        Лабытнанги                 0                    1           0   
766  Гусь-Хрустальный                 0                    1           1   
767           Бердяуш                 0                    1           0   
768            Лянтор                 0                    1           0   

     Проектная работа  
0                  24  
1                  13  
2                  14  
3                   7  
4                  13  
..                ...  
764                 0  
765                 0  
766                 0  
767                 0  
768                 0  

[769 rows x 5 columns]

In [55]:
connection.close()